### PR TITLE
separate `oldgametics` counters for floors and ceilings

### PR DESCRIPTION
**Pull request by rfomin**
Fix (?) this issue: [[doomworld]](https://www.doomworld.com/forum/topic/112333-this-is-woof-1120-jun-13-2023/?do=findComment&comment=2667488)

Not sure about this. Maybe we should go back to the previous approach?

### CODE REVIEW

**Comment by fabiangreffrath** [comment]
I think we need two separate `oldgametic` counters: one for floor and one for ceiling movement.

**Comment by rfomin** [comment]
> I think we need two separate `oldgametic` counters: one for floor and one for ceiling movement.

Done in https://github.com/fabiangreffrath/woof/pull/1120/commits/8234e55c019543d5484de8c7bb17ff1feaae4231. Seems to work so far.

File: src/r_bsp.c [hunk 1]
```diff
@@ -270,16 +270,18 @@ sector_t *R_FakeFlat(sector_t *sec, sector_t *tempsec,
...
-    if (uncapped &&
-        // Only if we moved the sector last tic.
-        sector->oldgametic == gametic - 1)
+    if (uncapped)
     {
         // Interpolate between current and last floor/ceiling position.
-        if (sector->floordata && sector->floorheight != sector->oldfloorheight)
+        if (sector->floordata && sector->floorheight != sector->oldfloorheight &&
+            // Only if we moved the sector last tic.
+            sector->oldfloorgametic == gametic - 1)
```
Comment (fabiangreffrath) [inline src/r_bsp.c:278]: Could you add braces here?

**Review by fabiangreffrath — APPROVED** [review]

**Comment by fabiangreffrath** [comment]
Thank you!

**Comment by MrAlaux** [comment]
Hold on, this doesn't seem to fix the Ancient Aliens Map24 lift I mentioned at Doomworld.

Here's a zip with a savegame right where the lift is, and the config file I used, just in case: [aaliens_map24_test.zip](https://github.com/fabiangreffrath/woof/files/11929032/aaliens_map24_test.zip)
The WAD should be the one available at idgames: [link](https://www.doomworld.com/idgames/levels/doom2/Ports/megawads/aaliens)

**Comment by rfomin** [comment]
> Hold on, this doesn't seem to fix the Ancient Aliens Map24 lift I mentioned at Doomworld.

This issue is present in Woof 10.5.1, seems to be old. Do you know the working version?

**Comment by MrAlaux** [comment]
> Do you know the working version?

Last time I played Ancient Aliens, at least that I remember, was in late 2021. I was most likely using an in-development version of Nugget Doom, between the 1.6.1 and 1.7.0 releases, which were based on Woof 8.1.0. However, I just tried the latter and the issue is still there.

Admittedly, I don't remember if the issue was present back then or not; I believe I was playing through a version of AA different to the one I'm using now. Regardless, the version of AA I'm using now, which should be the latest, definitely has the issue.

**Comment by MrAlaux** [comment]
I just remembered an identical issue in Dimension of The Boomed. Unlike the AA one, I remember coming across this one back then, although I don't remember what version of Woof/Nugget I was using.

Again, here's a zip with a savegame and config file: [dotb_map05_test.zip](https://github.com/fabiangreffrath/woof/files/11929155/dotb_map05_test.zip)
The WAD is available [here](https://www.moddb.com/mods/dimension-of-the-boomed/downloads/dimension-of-the-boomed-vanilla-boom).

**Comment by rfomin** [comment]
This patch fixes both aaliens.wad and dotb_rc3.wad:
```diff
diff --git a/src/r_bsp.c b/src/r_bsp.c
index 3da02d7..052798d 100644
--- a/src/r_bsp.c
+++ b/src/r_bsp.c
@@ -378,16 +378,16 @@ static void R_AddLine (seg_t *line)
   if (!backsector)
     goto clipsolid;
 
-  // killough 3/8/98, 4/4/98: hack for invisible ceilings / deep water
-  backsector = R_FakeFlat(backsector, &tempsec, NULL, NULL, true);
-
-  doorclosed = 0;       // killough 4/16/98
-
   // [AM] Interpolate sector movement before
   //      running clipping tests.  Frontsector
   //      should already be interpolated.
   R_MaybeInterpolateSector(backsector);
 
+  // killough 3/8/98, 4/4/98: hack for invisible ceilings / deep water
+  backsector = R_FakeFlat(backsector, &tempsec, NULL, NULL, true);
+
+  doorclosed = 0;       // killough 4/16/98
+
   // Closed door.
 
   if (backsector->interpceilingheight <= frontsector->interpfloorheight

```
Not sure if this correct. @fabiangreffrath What do you think?

**Comment by mikeday0** [comment]
I can reproduce the Eviternity issue. Sadly it's still present even with the changes from this PR. However rfomin's above patch fixes the problem.

**Comment by fabiangreffrath** [comment]
This is one of the "it's nice that it works, but I want to understand *why* it works" kind of fixes.

My first reaction is that it looks wrong to interpolate the backsector before replacing it with the fake backsector. On the other hand, if several different sectors share one fake backsector...

**Comment by fabiangreffrath** [comment]
Does moving only the `doorclosed = 0` line make any difference?

**Comment by rfomin** [comment]
> Does moving only the `doorclosed = 0` line make any difference?

No, it's not. Should I make a new PR?

**Comment by fabiangreffrath** [comment]
Yes, please. Don't get me wrong, I'm sure this code change is correct, but of like to investigate why it works before.

**Comment by rfomin** [comment]
> Don't get me wrong, I'm sure this code change is correct, but of like to investigate why it works before.

I don't know exactly how the "fake flats" work, but it looks like we're interpolating the "lower the drawn ceiling", hence the flickering.
I don't think this has worked correctly before, I checked Woof 6.0.0, this issue persists. I noticed problems in Dimension of the Boomed a long time ago, just forgot about them.

**Pull request merged**